<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
</head>

<body>
    <script>
        // async function test(){
        //     return new Promise((resolve,reject)=>{
        //         // resolve("ok")
        //         reject("fail")
        //     })
        // }
        // var a=test()
        // a.then(res=>{
        //     console.log(res);
        // }).catch(err=>{
        //     console.log(err);
        // })
        function ajax(url) {
            //用Promise封装ajax函数 
            //挂载在ajax上的缓存  先定义数值有缓存等于缓存 没有等于初始值
            var cache = ajax.cache || (ajax.cache = { data: null })
            if (cache.data) {
                console.log("缓存");
                // 有缓存 通过Promise.resolve将对象转化为promise 返回出去
                //返回一个Promise对象
                return Promise.resolve(cache.data)
            }
            // 没有缓存
            return new Promise((resolve, reject) => {
                var xhr = new XMLHttpRequest();
                xhr.open("get", url, true);
                xhr.send();
                xhr.onreadystatechange = function () {
                    if (xhr.readyState == 4) {
                        if (xhr.status >= 200 && xhr.status < 300) {
                            // 成功 返回值 在promise中返回就是promise
                            resolve(JSON.parse(xhr.responseText))
                            // 访问成功 缓存挂载在ajax上
                            ajax.cache.data = JSON.parse(xhr.responseText)
                        } else {
                            // 失败
                            reject(xhr.responseText);
                        }
                    }
                }
            })
        }
        async function getData() {
            var result = await ajax("http://xiongmaoyouxuan.com/api/tabs")
            return result
        }
        getData().then(res => {
            console.log("ok", res);
        }).catch(err => {
            console.log("err", err);
        })
    </script>
</body>

</html>